### DISCUSSION TITLE
Counting Asset<Mesh> for diagnostics/statistics pane

### DISCUSSION THREAD

**Discussion by johnny-smitherson** [bevyengine/bevy#13296] [Q&A]
I'm trying to figure out how many resources of each type there are inside `Res<Assets<X>>`, to display these in a diagnostics pane.

```rust

fn count_diagnostic(
    mut diagnostics: Diagnostics,
    ...
    asset_meshes: Res<Assets<Mesh>>,
    asset_stdmat: Res<Assets<StandardMaterial>>,
    asset_img: Res<Assets<Image>>,
) {
    ...
    diagnostics.add_measurement(&ASSET_MESH_COUNT, || asset_meshes.iter().count() as f64);
    diagnostics.add_measurement(&ASSET_STDMAT_COUNT, || asset_stdmat.iter().count() as f64);
    diagnostics.add_measurement(&ASSET_IMAGE_COUNT, || asset_img.iter().count() as f64);
}
```

This works great for Image and StandardMaterial, but for Mesh (which I create dynamically about 1300) the count is wrong. 

<img width="425" alt="Screenshot 2024-05-09 153540" src="https://github.com/bevyengine/bevy/assets/127537716/a400cf40-c2ca-463d-978a-fe2d84b99511">

It seems to only count the meshes that come from standard Bevy shapes:

```rust
    meshes.add(Sphere::new(radius).mesh().ico(4).unwrap());
```

But not counting the ones I create from scratch:

```rust
        let mut mesh = Mesh::new(
            PrimitiveTopology::TriangleList,
            RenderAssetUsages::RENDER_WORLD,
        );
        mesh.insert_indices(mesh::Indices::U32(all_indices));
        mesh.insert_attribute(Mesh::ATTRIBUTE_POSITION, all_verts);
        mesh.insert_attribute(Mesh::ATTRIBUTE_NORMAL, all_norms);
        mesh.insert_attribute(Mesh::ATTRIBUTE_UV_0, all_uvs);
        meshes.add(mesh);
```

All 1300 meshes are in the scene. If I count `Query<Handle<Mesh>>` the count is correct -- but that count gives back number of entities with a mesh, not number of distinct meshes, which is what I want. And if I computed the unique `Handle<Mesh>` inside the system, then the resulting count wouldn't include meshes that are not inserted on any Entities but are still held somewhere (e.g. inside another resource).

Any idea what's happening wrong with this count? Is this a bug, or a feature?

**Comment by mockersf** [comment]
meshes that are marked as `RenderAssetUsages::RENDER_WORLD` are removed from storage once uploaded to the GPU, to be able to free memory
